### PR TITLE
Add an advanced subscription block.

### LINKED ISSUE #62: Create an advanced subscriptions block
I have a current feature request wherein the site wants a user to be able to subscribe to all available subscriptions on a given page.

For instance, take an article. There is the subscription to the article itself, but also, the author, the various taxonomy terms referenced, the content type itself, etc.

I think adding a block to the message_subscribe_ui module would be relatively straightforward in this regards.

### PR DESCRIPTION
**Pull request by jhedstrom**
- Fixes #62

### CODE REVIEW

**Comment by amitaibu** [comment]
Can you attach a screenshot of how it looks please.

**Comment by jhedstrom** [comment]
Fails should have been due to https://www.drupal.org/node/2827864.

I'll attach a screen shot when I get closer.

**Comment by jhedstrom** [comment]
The requested screenshot.
screen shot 2016-11-16 at 1 22 19 pm

**Comment by jhedstrom** [comment]
Note those labels for the subscriptions are fairly generic. We could be more prescriptive/assertive here, and do something like this:

_Subscribe_
- [ ] To this [content type]
- [ ] Content tagged with [taxonomy term]
- [ ] Content by author [author name]
- [ ] All content in [content type] 

Re-using the flag label seems tricky here, since by default it would just be `Subscribe` next to a checkbox :)

File: message_subscribe_ui/src/Plugin/Block/Subscriptions.php [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    return 'message_subscribe_ui_block';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $form = [];
+    $subscribable = FALSE;
+    if ($entity = $this->getCurrentEntity()) {
```
Comment (amitaibu) [inline message_subscribe_ui/src/Plugin/Block/Subscriptions.php]: I usually prefer to retrun early. So

``` php
if (!$entity = $this->getCurrentEntity()) {
  return [];
}
```

File: message_subscribe_ui/src/Plugin/Block/Subscriptions.php [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+          '#markup' => $this->t('Manage all <a href=":url">subscriptions</a>.', [':url' => Url::fromRoute('message_subscribe_ui.tab', ['user' => \Drupal::currentUser()->id()])->toString()]),
+        ];
+        $form['subscriptions'] = [
+          '#type' => 'container',
+          '#tree' => TRUE,
+        ];
+        foreach ($entities as $referenced_entity) {
+          $flags = $this->subscribers->getFlags($referenced_entity->getEntityTypeId(), $referenced_entity->bundle(), $this->currentUser);
+          if (!empty($flags)) {
+            // @todo Support multiple subscription flags per-entity if there is
```
Comment (amitaibu) [inline message_subscribe_ui/src/Plugin/Block/Subscriptions.php]: what do you mean by this?

Comment (jhedstrom) [reply message_subscribe_ui/src/Plugin/Block/Subscriptions.php]: Well, the `getFlags` method returns an array, and here I just grab one as I assume there won't be more than one subscribe flag per entity type. However, there _could_ be in theory anyway, thus the `@todo`. I can replace the `@todo` with a simpler note explaining this logic...

**Comment by jhedstrom** [comment]
If this doesn't make sense here, I'm also fine moving it to a separate module for now. Message subscribe UI has always been about configuring via flags, so this would be a departure from that in terms of how the labels are working in this block.

**Comment by amitaibu** [comment]
> so this would be a departure from that in terms of how the labels are working in this block.

Looking at the code I see you are still using flag to control the subscription, so it the UI is a checkbox it's still fine - as long as the underlaying data is the same

**Comment by jhedstrom** [comment]
I need to add in access checking, then this should be good to go.

**Comment by jhedstrom** [comment]
Access check is in place, and I also refactored the label logic into a helper method. Folks can readily tweak those labels via a form alter hook.

**Pull request merged**